tools/chpldef/Message.cpp

```diff
@@ -0,0 +1,294 @@
+      return nullptr;
+    } else {
+      params = ptr;
+    }
+  }
+
+  switch (tag) {
+    #define CHPLDEF_MESSAGE(name__, x1__, x2__, x3__) \
+    case name__: return name__::create(std::move(id), *params);
+    #include "./message-macro-list.h"
+    #undef CHPLDEF_MESSAGE
+    default: break;
+  }
+
+  CHPLDEF_IMPOSSIBLE();
+
+  return nullptr;
+}
+
+// TODO: Is outbound format the same as for incoming requests?
+template <typename P, typename R>
+JsonValue Request<P, R>::pack() const {
+  return p.toJson();
+}
+
+template <typename T>
+static Response createSuccessfulResponse(const T* t) {
+  CHPL_ASSERT(t && t->isCompleted() && !t->hasError());
+  auto r = t->result();
+  CHPL_ASSERT(r);
+  auto data = r->toJson();
+  auto ret = Response::create(t->id(), std::move(data));
+  return ret;
+}
+
+opt<Response> Message::response(Server* ctx, const Message* msg) {
+  opt<Response> ret;
+
+  if (msg->tag() == Message::UNSET || msg->tag() == Message::INVALID ||
+      msg->tag() == Message::RESPONSE) {
+    return ret;
+  }
+
+  switch (msg->status()) {
+    case COMPLETED: {
+      CHPL_ASSERT(msg->error() == Message::OK);
+      switch (msg->tag()) {
+        #define CHPLDEF_MESSAGE(name__, x1__, x2__, x3__) \
+        case name__: return createSuccessfulResponse(msg->to##name__());
+        #include "./message-macro-list.h"
+        #undef CHPLDEF_MESSAGE
+        default: break;
+      }
+    } break;
+    case FAILED: {
+      CHPL_ASSERT(msg->error() != Message::OK);
+      // TODO: What debug JSON to embed here, if any?
+      JsonValue data = nullptr;
+      ret = Response::create(msg->id(), msg->error(), msg->note(),
+                             std::move(data));
+    } break;
+    default: break;
+  }
+
+  return ret;
+}
+
+bool Message::isOutbound() const {
+  if (tag_ == Message::RESPONSE) return true;
+  #define CHPLDEF_MESSAGE(name__, outbound__, x2__, x3__) \
+  if (tag_ == Message::name__) return outbound__;
+  #include "./message-macro-list.h"
+  #undef CHPLDEF_MESSAGE
+  return false;
+}
+
+bool Message::isNotification() const {
+  if (tag_ == Message::RESPONSE) return true;
+  #define CHPLDEF_MESSAGE(name__, x1__, notification__, x3__) \
+  if (tag_ == Message::name__) return notification__;
+  #include "./message-macro-list.h"
+  #undef CHPLDEF_MESSAGE
+  return false;
+}
+
+const char* Message::jsonRpcMethodName() const {
+  return tagToJsonRpcMethodName(tag_);
+}
+
+const char* Message::tagToJsonRpcMethodName(Message::Tag tag) {
+  switch (tag) {
+    #define CHPLDEF_MESSAGE(name__, x1__, x2__, rpc__) \
+    case name__: return #rpc__;
+    #include "./message-macro-list.h"
+    #undef CHPLDEF_MESSAGE
+    default: break;
+  }
+
+  return nullptr;
+}
+
+Message::Tag Message::jsonRpcMethodNameToTag(const char* str) {
+  #define CHPLDEF_MESSAGE(name__, x1__, x2__, rpc__) \
+  if (!strcmp(#rpc__, str)) return name__;
+  #include "./message-macro-list.h"
+  #undef CHPLDEF_MESSAGE
+  return Message::INVALID;
+}
```

Comment:
An optimization opportunity is to use a [Trie](https://en.wikipedia.org/wiki/Trie) here.

Reply:
I put a comment.